Replay Settings

EVAST9919 opened the pull request:
![screenshot-120](https://cloud.githubusercontent.com/assets/22874522/26524589/a724c2b4-4340-11e7-9be1-8b70d4ba54d7.png)

EVAST9919 committed:
Moved music controller buttons to it's own class

EVAST9919 committed:
ingame options basic logic

EVAST9919 committed:
Moved collectionsDropdown to it's own class

EVAST9919 committed:
More overlays and overall fixes

EVAST9919 committed:
Visible only in replay

EVAST9919 committed:
Added bindable to adjust playback speed

EVAST9919 committed:
Merge remote-tracking branch 'refs/remotes/ppy/master' into ingame_options

EVAST9919 committed:
fixed license header

EVAST9919 committed:
Merge remote-tracking branch 'refs/remotes/ppy/master' into ingame_options

EVAST9919 committed:
CI fixes

EVAST9919 committed:
now button can hide content

EVAST9919 committed:
Merge remote-tracking branch 'refs/remotes/ppy/master' into ingame_options

EVAST9919 committed:
Merge remote-tracking branch 'refs/remotes/ppy/master' into ingame_options

EVAST9919 committed:
better resizing

EVAST9919 committed:
options -> settings

EVAST9919 committed:
Merge remote-tracking branch 'refs/remotes/ppy/master' into ingame_options

EVAST9919 committed:
Make field readonly

EVAST9919 committed:
Merge remote-tracking branch 'refs/remotes/ppy/master' into ingame_options

EVAST9919 committed:
Naming fixes

EVAST9919 committed:
dropdown-like animation

EVAST9919 committed:
More constants

EVAST9919 committed:
Change visibility on ctrl+h

EVAST9919 committed:
removed forgotten file

EVAST9919 committed:
Merge remote-tracking branch 'refs/remotes/ppy/master' into ingame_options

EVAST9919 committed:
Fixed the ability to change visibility even not in replay

EVAST9919 committed:
moved autosize to ctor and whitespace fix

EVAST9919 committed:
Extended testcase

EVAST9919 committed:
Merge remote-tracking branch 'refs/remotes/ppy/master' into ingame_options

EVAST9919 committed:
Merge remote-tracking branch 'refs/remotes/ppy/master' into ingame_options

EVAST9919 committed:
SettingsContainer -> SettingsDropdownContainer

EVAST9919 committed:
Settings -> ReplaySettings

EVAST9919 committed:
Merge remote-tracking branch 'refs/remotes/ppy/master' into ingame_options

EVAST9919 committed:
oops

EVAST9919 committed:
Merge remote-tracking branch 'refs/remotes/ppy/master' into ingame_options

EVAST9919 committed:
overall fixes

EVAST9919 committed:
Merge remote-tracking branch 'refs/remotes/ppy/master' into ingame_options

EVAST9919 committed:
Toogle visibility without ctrl pressed

EVAST9919 committed:
More xmldoc and few Title text fixes

smoogipoo committed:
Merge branch 'master' into ingame_options

smoogipooo commented on osu.Game/Screens/Play/ReplaySettings/ReplaySettingsOverlay.cs:
```diff
@@ -0,0 +1,73 @@
+        /// <summary>
+        /// Allow user to interact with this overlay.
+        /// </summary>
+        public bool IsAvaliable
```
How about doing `ReplaySettingsOverlay.Hide()` from inside `HUDOverlay.BindHitRenderer()`? Similar to how ModDisplay is done. Would remove this property + tidy up any usages.

Would also do the same for the key bind I think, further removing toggleVisibility, keeping this a pretty basic class.

smoogipooo commented on osu.Game/Graphics/UserInterface/SettingsDropdownContainer.cs:
```diff
@@ -0,0 +1,126 @@
+                            RelativeSizeAxes = Axes.X,
+                            Height = header_height,
+                            Origin = Anchor.TopCentre,
+                            Anchor = Anchor.TopCentre,
```
Prefer Origins + Anchors at the top of the initializer for consistency (there are other cases in this PR where this is not the case).

smoogipooo commented on osu.Game/Graphics/UserInterface/SettingsDropdownContainer.cs:
```diff
@@ -0,0 +1,126 @@
+            };
+        }
+
+        public new void Add(Drawable drawable)
```
`protected override Container<Drawable> Content => content`, change `Children =` to `InternalChildren =` and then this method can be removed.

smoogipooo commented on osu.Game/Graphics/UserInterface/SimpleButton.cs:
```diff
@@ -0,0 +1,114 @@
+                    Size = new Vector2(button_size),
+                    CornerRadius = 5,
+                    Masking = true,
+
```
No space.

smoogipooo commented on osu.Game/Screens/Play/ReplaySettings/ReplaySettingsOverlay.cs:
```diff
@@ -0,0 +1,73 @@
+            return base.OnKeyDown(state, args);
+        }
+
+        private void toogleVisibility()
```
`toggleVisibility`

smoogipooo commented on osu.Game/Screens/Play/ReplaySettings/CollectionSettings.cs:
```diff
@@ -0,0 +1,31 @@
+        [BackgroundDependencyLoader]
+        private void load()
+        {
+            Add(new OsuSpriteText
```
With my suggestion of `Content => content` in the other review comment, you can tidy this up using
`Children = new Drawable[] { ... };`

smoogipoo reviewed (changes requested):
First code-only pass.

EVAST9919 committed:
Merge remote-tracking branch 'refs/remotes/ppy/master' into ingame_options

EVAST9919 committed:
Applied suggested changes

smoogipoo committed:
Allow Nub/OsuCheckbox/OsuSliderBar colours to be overridden.

smoogipoo committed:
Make replay settings match design colours.

smoogipooo commented:
I think we want to leave this in visual tests for now, until some of these options are implemented, then enable them selectively. We're trying not to have options in the game which aren't implemented yet.

smoogipooo commented on osu.Game/Screens/Play/ReplaySettings/DiscussionSettings.cs:
```diff
@@ -0,0 +1,36 @@
+            {
+                new ReplaySettingsCheckbox
+                {
+                    LabelText = "Show floating coments",
```
comments*

smoogipooo commented on osu.Game/Graphics/UserInterface/SettingsDropdownContainer.cs:
```diff
@@ -0,0 +1,123 @@
+                    Direction = FillDirection.Vertical,
+                    RelativeSizeAxes = Axes.X,
+                    AutoSizeAxes = Axes.Y,
+
```
No whitespace in object initializer.

smoogipooo commented on osu.Game/Graphics/UserInterface/SettingsDropdownContainer.cs:
```diff
@@ -0,0 +1,123 @@
+
+        private FillFlowContainer content;
+        private SimpleButton button;
+        private bool buttonIsActive;
```
buttonIsActive -> expanded

smoogipooo commented on osu.Game/Graphics/UserInterface/SimpleButton.cs:
```diff
@@ -0,0 +1,114 @@
+
+namespace osu.Game.Graphics.UserInterface
+{
+    public class SimpleButton : ClickableContainer
```
Consider renaming this to IconButton.

smoogipooo commented on osu.Game/Screens/Play/HUDOverlay.cs:
```diff
@@ -95,8 +99,13 @@ public virtual void BindHitRenderer(HitRenderer hitRenderer)
             if (!hitRenderer.HasReplayLoaded)
+            {
+                ReplaySettingsOverlay.Hide();
+                ReplaySettingsOverlay.AlwaysPresent = false;
```
What is the AlwaysPresent value here for?

EVAST9919 replied:
We need to block toggling visibility if we are not in replay

smoogipooo replied:
That's why I suggested putting the key event for toggling visibility in this class' OnKeyDown.

smoogipooo commented on osu.Game/Graphics/UserInterface/SettingsDropdownContainer.cs:
```diff
@@ -0,0 +1,123 @@
+using osu.Framework.Graphics.Sprites;
+using osu.Game.Graphics.Sprites;
+
+namespace osu.Game.Graphics.UserInterface
```
I think it's best if this resided inside Screens/Play/ReplaySettings for now.

smoogipooo commented on osu.Game/Graphics/UserInterface/SettingsDropdownContainer.cs:
```diff
@@ -0,0 +1,123 @@
+
+namespace osu.Game.Graphics.UserInterface
+{
+    public abstract class SettingsDropdownContainer : Container
```
ReplaySettingsGroup.

smoogipooo commented on osu.Game/Graphics/UserInterface/SettingsDropdownContainer.cs:
```diff
@@ -0,0 +1,123 @@
+    public abstract class SettingsDropdownContainer : Container
+    {
+        /// <summary>
+        /// The title of this container, which will be written in header.
```
If class is renamed to ReplaySettingsGroup -> "The title to be displayed in the header of this group."

smoogipooo commented on osu.Game/Graphics/UserInterface/SettingsDropdownContainer.cs:
```diff
@@ -0,0 +1,123 @@
+
+        [BackgroundDependencyLoader]
+        private void load(OsuColour colours)
+        {
```
All of this method should be moved to the constructor, after which `content` and `button` can be made readonly. Use load() for initializing the button colour.

Generally I think we should keep ctors for object initialization and load() for virtual methods + DI initialization. Having this stuff inside load() messes with object initializers because the load() method would override them - possibly unexpectedly.

smoogipooo replied:
And make sure to move every other class from load() to ctor along with this.

smoogipooo commented on osu.Game/Graphics/UserInterface/SettingsDropdownContainer.cs:
```diff
@@ -0,0 +1,123 @@
+                            Anchor = Anchor.TopCentre,
+                            RelativeSizeAxes = Axes.X,
+                            Height = header_height,
+
```
And here.

smoogipooo commented on osu.Game/Screens/Play/ReplaySettings/CollectionSettings.cs:
```diff
@@ -0,0 +1,34 @@
+                {
+                    Text = @"Add current song to",
+                },
+                new CollectionsDropdown<PlaylistCollection>
```
We'll need `CollectionsDropdown : SettingsDropdown` to provide the label for us, otherwise the label and the dropdown are too far apart from one another. Check how I handled  this with ReplaySettingsCheckbox/ReplaySettingsSliderBar.

smoogipoo reviewed (dismissed):
See comments.

smoogipoo committed:
Fix usings.

EVAST9919 committed:
Merge remote-tracking branch 'refs/remotes/ppy/master' into ingame_options

smoogipooo commented on osu.Game/Overlays/Settings/SettingsDropdown.cs:
```diff
@@ -27,7 +27,7 @@ public class SettingsDropdown<T> : SettingsItem<T>
         }
 
-        protected override Drawable CreateControl() => dropdown = new OsuDropdown<T>
+        protected override Drawable CreateControl() => dropdown = new Graphics.UserInterface.SettingsDropdown<T>
```
Is this change intended?

smoogipooo commented on osu.Game/Graphics/UserInterface/OsuDropdown.cs:
```diff
@@ -13,7 +13,7 @@
 namespace osu.Game.Graphics.UserInterface
 {
-    public class OsuDropdown<T> : Dropdown<T>
+    public class SettingsDropdown<T> : Dropdown<T>
```
Guess that goes along with this, I don't think this change is intended :P

EVAST9919 committed:
Applied suggested changes

EVAST9919 committed:
CI fixes

EVAST9919 committed:
Merge branch 'master' of https://github.com/ppy/osu into ingame_options

peppy committed:
Merge remote-tracking branch 'upstream/master' into ingame_options

peppy commented:
https://streamable.com/ul5dk

This feels really wrong

peppy commented on osu.Game/Screens/Play/ReplaySettings/ReplaySettingsGroup.cs:
```diff
@@ -0,0 +1,127 @@
+
+        private readonly FillFlowContainer content;
+        private readonly IconButton button;
+        private bool expanded;
```
this is always opposite of what is should be...

peppy committed:
Fix lots of small issues

peppy committed:
Fix namespaces and class names

peppy committed:
Don't display in-game

Nothing does anything yet, so we can't display this. Make it functional before adding to game project.

peppy commented:
I'm going to merge this with my changes. There needs to be changes to the underlying `FillFlow` implementation to fix the animation issues.

I've stopped this from being displayed outside of the VisualTest as it adds no functionality to the game. It can be added once it has functionality.

peppy committed:
Remove bad
